File: Leetcode/560/560.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int n = (int)nums.size();
        vector<int> pre(n+1);
        for (int i = 1; i <= n; ++i) {
            pre[i] = pre[i-1] + nums[i-1];
        }
        int counter = 0;
        // Case 1:
        for (int i = 1; i <= n; ++i) {
            if (pre[i] == k) {
                ++counter;
            }
        }
        // Case 2:
        unordered_map<int, vector<int> > unmap;
        for (int i = 1; i <= n; ++i) {
            unmap[pre[i]].emplace_back(i);
        }
        for (int l = 1; l <= n; ++l) {
            if (unmap.find(pre[l]+k) != unmap.end()) {
                auto &v = unmap[pre[l]+k];
                auto iter = upper_bound(v.begin(),v.end(),l);
                if (iter != v.end()) {
                    counter += (int)(v.end() - iter);
                }
            }
        }
        return counter;
    }
};

File: Leetcode/560/560.go
func subarraySum(nums []int, k int) int {
	// pre[r] == pre[l-1] + k
	pre := make([]int, len(nums)+1)
	for i := 1; i <= len(nums); i++ {
		pre[i] = pre[i-1] + nums[i-1]
	}
	unmap := make(map[int][]int)
	counter := 0
	for i := 0; i <= len(nums); i++ {
		unmap[pre[i]] = append(unmap[pre[i]], i)
	}
	for i := 0; i <= len(nums); i++ {
		val, ok := unmap[pre[i]+k]
		if ok {
			temp := sort.Search(len(val), func(j int) bool {
				return val[j] > i
			})
			counter += len(val) - temp
		}
	}
	return counter
}